lib/services/access_checker.go

```diff
@@ -217,6 +241,9 @@ type accessChecker struct {
 	// to search-based access requests) will be implemented by
 	// accessChecker.
 	RoleSet
+
+	// PredicateAccessChecker is embedded to allow access checking via access policy resources.
+	*predicate.PredicateAccessChecker
```

Comment:
does it need to be embedded? RoleSet was embedded to avoid a large refactor when switching from RoleSet to AccessChecker everywhere. I think it might be preferable to avoid the embedding so that you can't accidentally call a PredicateAccessChecker method without considering the RoleSet. The same argument could be made for not embedding RoleSet now that there are two access deciders, you don't want to accidentally ignore the PredicateAccessChecker

Reply:
This is something that will be resolved in https://github.com/gravitational/teleport/pull/20556. It's prohibitive to do here due to how the the accesschecker interface needs to be reworked.